### ISSUE TITLE
UNION query unable to resolve correct bind and fails with "(sqlite3.OperationalError) no such table"

### ISSUE THREAD

**Issue by alexjironkin** [pallets-eco/flask-sqlalchemy#1231]
This is related to #1210 and in general about finding correct bind to use for execution.

We were looking to upgrade to v3 to take advantage of the dynamic bind changes we need to and and came across this problem: `UNION` queries (maybe others too) are falling back to using default bind (`None` key in `SQLALCHEMY_BINDS`). Below is sample testcase that reproduces the problem:

```python
from flask import Flask
from flask_sqlalchemy import SQLAlchemy
import sqlalchemy as sa
from sqlalchemy import func, union


def test_union_query() :
    app = Flask("test-app")
    app.config["SQLALCHEMY_DATABASE_URI"] = "sqlite:///"
    app.config["SQLALCHEMY_BINDS"] = {"db1": "sqlite:///"}

    db = SQLAlchemy(app)

    class UserType1(db.Model):
        __bind_key__ = "db1"
        __tablename__ = "user_type_1"
        id = sa.Column(sa.Integer, primary_key=True)
        user_type = sa.Column(sa.Integer, nullable=False)
        name = sa.Column(sa.String(50), nullable=False)

    class UserType2(db.Model):
        __bind_key__ = "db1"
        __tablename__ = "user_type_2"
        id = sa.Column(sa.Integer, primary_key=True)
        user_type = sa.Column(sa.Integer, nullable=False)
        name = sa.Column(sa.String(50), nullable=False)

    with app.app_context():
        db.create_all()

        db.session.add(UserType1(id=1, user_type=1, name="alice"))
        db.session.add(UserType1(id=2, user_type=1, name="bob"))
        db.session.add(UserType2(id=2, user_type=2, name="charlie"))

        query_1 = db.session.query(
            func.count(UserType1.name).label("count"),
            UserType1.user_type.label("user_type"),
        ).group_by(UserType1.user_type)

        assert db.session.query(query_1.subquery()).all()
        query_2 = db.session.query(
            func.count(UserType2.name).label("count"),
            UserType2.user_type.label("user_type"),
        ).group_by(UserType2.user_type)
        assert db.session.query(query_2.subquery()).all()

        union_query = union(query_1, query_2)

        assert len(db.session.query(union_query.subquery()).all()) == 2
```
The individual queries work as expected, however as soon as we do `union(query_1, query_2)` we get a nasty error:

`FAILED tests/test_union.py::test_union_query - sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) no such table: user_type_2`

Debugging this further it appears that `Session.get_bind` is called only with `clause` argument and `mapper` is `None`. The `clause` is some selectable: `<sqlalchemy.sql.selectable.Select object at 0x7f260bf6a5e0>` but this doesn't have `metadata`, `bind` or `table` attributes. 

Looking further still, this seems to be an issue with SqlAlchemy 1.4.x in general (I tested with 1.4.0, 1.4.10, 1.4.18, 1.4.19, 1.4.19 and 1.4.48), and upgrading to SQLAlchemy 2.x (2.0.17 at the time of writing) seem to resolve issue and `mapper`  is now passed in to `Session.get_bind` as `<Mapper at 0x7fbc1ba1ecd0; UserType1>`. So my suspicion is that for a (small) subset of queries like union, flask_sqlalchemy is not compatible with sqlalchemy 1.4.x and requires version 2. 

I was not able to come up with a detection mechanism for `clause` that would resolve to a table in this instance, so that a bind can be found. Any thoughts? I am also going to file this with sqlalchemy in hope they can patch to 1.4.49 to have this fix in as unfortunately we have another dependency with hard pin `sqlalchemy<2` so upgrading is not currently an option.

Do you want me to create a PR with the test case above? 

Environment:

- Python version: 3.9.16
- Flask-SQLAlchemy version: 3.0.5
- SQLAlchemy version: 1.4.48

**Comment by alexjironkin** [comment]
Even more troubleshooting reveals that this was fixed in sqlalchemy 2.0.15 release, in particular this looks relevant: 

https://github.com/sqlalchemy/sqlalchemy/issues/9805

as `_propagate_attrs` has the information mapper needed. At least looking at the internals of the `Session` before `get_bind` is called.

**Comment by alexjironkin** [comment]
- This is not going to be fixed in 1.4.x (see https://github.com/sqlalchemy/sqlalchemy/issues/10043) and upgrading to 2.0.15+ is the preferred way of addressing this.
- If you are not able to upgrade for whatever reason and need to stay on 1.4.x @zzzeek has given a workaround to modify `_propagate_attrs` in the following way (using example above):

```python
from sqlalchemy import inspect

def test_union_query():
        # as before
        union_query = db.session.query(union(query_1, query_2))
        union_query ._propagate_attrs = union_query ._propagate_attrs.union(
            {"plugin_subject": sa.inspect(UserType1), "compile_state_plugin": "orm"}
        )
```

Then `mapper` is correctly passed to `Session.get_bind`. If you have a complex query, I am not sure if it matters which model you pick for matter. It's probably not going across binds, so does it matter?
